sympy/combinatorics/fp_groups.py

```diff
@@ -0,0 +1,369 @@
+    >>> c
+    [[0, 0, 1, 2], [-1, -1, 2, 0], [-1, -1, 0, 1]]
+    >>> scan(c, 0, x**-1*y**-1*x*y, A)
+    >>> c
+    [[0, 0, 1, 2], [-1, -1, 2, 0], [2, 2, 0, 1]]
+    >>> scan(c, 1, x**-1*y**-1*x*y, A)
+    >>> c
+    [[0, 0, 1, 2], [1, 1, 2, 0], [2, 2, 0, 1]]
+
+    # Example 5.2, Pg 154
+    >>> c1 = [[1, 1, -1, -1], [0, 0, -1, -1]]
+    >>> scan(c1, 0, x*y, A)
+    >>> c1
+    [[1, 1, -1, 1], [0, 0, 0, -1]]
+
+    >>> c2 = [[1, 1, 2, 1], [0, 0, 0, -1], [-1, -1, -1, 0]]
+    >>> scan(c2, 1, y**3, A)
+    >>> c2
+    [[1, 1, 2, 1], [0, 0, 0, 2], [-1, -1, 1, 0]]
+
+    >>> c3 = [[1, 1, 2, 1], [0, 0, 0, 2], [3, 3, 1, 0], [2, 2, -1, -1]]
+    >>> scan(c3, 2, (x*y)**3, A)
+    >>> c3
+    [[1, 1, 2, 1], [0, 0, 0, 2], [3, 3, 1, 0], [2, 2, 3, 3]]
+
+    """
+    # alpha is an integer representing a "coset"
+    # since scanning can be in two cases
+    # 1. for alpha=0 and w in Y (i.e generating set of H)
+    # 2. alpha in sigma (set of live cosets), w in R (relators)
+    f = alpha
+    i = 0
+    r = len(word)
+    # list of union of generators and their inverses
+    # A = coset_table.A
+    while i < r and C[f][A.index(word.subword(i, i+1))] != -1:
+        f = C[f][A.index(word.subword(i, i+1))]
+        i += 1
+    if i >= r:
+        if f != alpha:
+            # implement the "coincidence" routine on Pg 158 of Handbook.
+            coincidence(C, f, alpha)
+            return
+    b = alpha
+    j = r
+    while j >= i and C[b][A.index(word.subword(j-1, j).inverse())] != -1:
+        b = C[b][A.index(word.subword(j-1, j).inverse())]
+        j -= 1
+    if j < i:
+        # we have an incorrect completed scan with coincidence f ~ b
+        # run the "coincidence" routine
+        # line: We assume that before each call of COINCIDENCE, for α ∈ [1..n],
+        # we have p[α]=α iff α ∈ Ω.
+        p.extend(list(range(len(C))))
+        p[b] = f
+        coincidence(C, f, b)
+    elif j == i + 1:
+        # deduction process
+        C[f][A.index(word.subword(i, i+1))] = b
+        C[b][A.index(word.subword(i, i+1).inverse())] = f
+    # otherwise scan is incomplete and yields no information
+
+
+# here alpha, beta represent the pair of cosets where
+# coicidence occurs
+
+def merge(k, lamda, p, q, l):
+    phi = rep(k, p)
+    chi = rep(lamda, p)
+    if phi != chi:
+        mu = min(phi, chi)
+        v = max(phi, chi)
+        p[v] = mu
+        l += 1
+        q[l] = v
+    return l
+
+
+def rep(k, p):
+    lamda = k
+    pho = p[lamda]
+    while p != lamda:
+        lamda = pho
+        pho = p[lamda]
+    mu = k
+    pho = p[mu]
+    while pho != lamda:
+        p[mu] = lamda
+        mu = pho
+        pho = p[mu]
+    return lamda
+
+
+# alpha, beta coincide
+def coincidence(C, alpha, beta):
+    l = 0
+    q = []
+    #XXX what are p, q here?
+    l = merge(alpha, beta, p, q, l)
+    i = 1
+    while i <= l:
+        gamma = q[i]
+        i += 1
+        del C[gamma]
```

Comment:
It seems that this is already implicitly done by `merge` when setting `p[v] = mu`.